File: frontend/src/api/substitutions.ts
// 調代課處理:代課推薦、指派處置(M4-2)、可對調節次(v1.2.1)、通知單(v1.2.2/v1.2.5)。

import { apiDelete, apiGet, apiPut } from '@/api/client'

export interface Candidate {
  teacher_id: number
  teacher_name: string
  same_subject: boolean
  at_school_that_day: boolean
  sub_periods_this_month: number
  reasons: string[]
}

export interface Recommendation {
  affected_period_id: number
  candidates: Candidate[]
  no_candidate_hint: string
}

export interface Substitution {
  id: number
  affected_period_id: number
  type: string
  type_label: string
  handler_teacher_id: number | null
  handler_name: string | null
  counts_toward_hours: boolean
  funding_source: string
  swap_date: string | null
  swap_period_name: string
  swap_class_names: string
  swap_subject_name: string
  created_by_name: string
}

/** 乙的某一節課、在某一天,可改由請假的甲來補 */
export interface SwapOption {
  entry_id: number
  date: string
  weekday: number
  period_no: number
  period_name: string
  class_names: string
  subject_name: string
  same_class: boolean // 與請假那節同一班(同班互調)
  in_block: boolean // 連堂中的一節(只換這一節)
}

export interface SwapPartner {
  teacher_id: number
  teacher_name: string
  teaches_same_class: boolean
  blocked_reason: string // 非空 = 這位老師在請假那節就來不了
  options: SwapOption[]
}

export interface SwapOptions {
  affected_period_id: number
  date_from: string
  date_to: string
  partners: SwapPartner[]
}

export interface AssignBody {
  type: string
  handler_teacher_id?: number | null
  counts_toward_hours?: boolean | null
  funding_source?: string
  swap_entry_id?: number | null
  swap_date?: string | null
  swap_period_no?: number | null // 連堂格位換其中哪一節
}

export const listSubstitutionTypes = (): Promise<Record<string, string>> =>
  apiGet('/substitution-types')

export const getRecommendations = (affectedId: number): Promise<Recommendation> =>
  apiGet(`/affected-periods/${affectedId}/recommendations`)

export const assignSubstitution = (affectedId: number, body: AssignBody): Promise<Substitution> =>
  apiPut(`/affected-periods/${affectedId}/substitution`, body)

export const clearSubstitution = (affectedId: number): Promise<{ status: string }> =>
  apiDelete(`/affected-periods/${affectedId}/substitution`)

/** 未指定教師時,只找也教這個班的老師;weeks = 從請假那週起找幾週(1–4) */
export const getSwapOptions = (
  affectedId: number, teacherId?: number | null, weeks = 2,
): Promise<SwapOptions> => {
  const q = new URLSearchParams({ weeks: String(weeks) })
  if (teacherId) q.set('teacher_id', String(teacherId))
  return apiGet(`/affected-periods/${affectedId}/swap-options?${q}`)
}

// ── 通知單(調課單 v1.2.2、代課單 v1.2.5;兩者版面相同,表頭與格子內容不同)──

export interface SlipRow {
  ordinal: number // 第幾節(只數一般課)
  start_time: string | null
  end_time: string | null
  afternoon_starts: boolean // 下午第一節:紙本在這裡畫粗線
}

export interface SlipCell {
  date: string
  weekday: number
  ordinal: number
  subject_name: string
  actor: string // 第三行:調課單寫上課老師;代課單寫「班級[代]」或「老師[代]」
  code: string // 「調09-15_25」:與 9/15 星期二第 5 節對調;代課單沒有
}

export interface SlipWeek {
  monday: string
  days: string[]
  cells: SlipCell[]
}

export interface Slip {
  kind: 'teacher' | 'class'
  teacher_name: string
  class_names: string
  date_from: string
  date_to: string
  absent_teacher_name: string // 代課單:請假教師
  leave_type_name: string // 代課單:假別
  funding_label: string // 代課單:計費方式
  rows: SlipRow[]
  weeks: SlipWeek[]
}

export interface Slips {
  title: string
  kind: 'swap' | 'substitute'
  slips: Slip[]
}

const slipsQuery = (semesterId: number, affectedIds: number[]): string => {
  const q = new URLSearchParams({ semester_id: String(semesterId) })
  for (const id of affectedIds) q.append('affected_period_ids', String(id))
  return String(q)
}

export const getSwapSlips = (semesterId: number, affectedIds: number[]): Promise<Slips> =>
  apiGet(`/swap-slips?${slipsQuery(semesterId, affectedIds)}`)

export const getSubstituteSlips = (semesterId: number, affectedIds: number[]): Promise<Slips> =>
  apiGet(`/substitute-slips?${slipsQuery(semesterId, affectedIds)}`)

export const listFundingSources = (): Promise<string[]> =>
  apiGet('/substitution-funding-sources')

/** 在新分頁開啟通知單(列印頁不套側邊欄);kind 決定印調課單還是代課單 */
export function openSlips(
  kind: 'swap' | 'substitute', semesterId: number, affectedIds: number[],
): void {
  const path = kind === 'swap' ? '/swap-slips/print' : '/substitute-slips/print'
  window.open(`${path}?semester_id=${semesterId}&ids=${affectedIds.join(',')}`, '_blank')
}

File: backend/app/api/substitutions.py
"""調代課處理工作台:代課推薦、指派處置、調課(M4-2)。

處置是行政決定,一律限教學組長/教務主任。教師端不在這裡處理(只在通知端「確認收到」)。
"""

from dataclasses import asdict

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.auth import require_roles
from app.core.db import get_db
from app.models.audit import AuditLog
from app.models.leave import AffectedPeriod
from app.models.substitution import (
    FUNDING_SOURCES,
    SUBSTITUTION_TYPE_CN,
    Substitution,
    SubstitutionType,
)
from app.models.user import Role, User
from app.schemas.substitution import (
    AssignRequest,
    CandidateOut,
    RecommendationOut,
    SlipsOut,
    SubstitutionOut,
    SwapOptionsOut,
)
from app.services import sub_slips, swap_options, swap_slips
from app.services import substitution_recommender as recommender
from app.services import substitutions as sub_service
from app.services.slip_layout import SlipError

router = APIRouter(tags=["substitutions"])

editor = require_roles(Role.scheduler, Role.director)


def _get_affected(db: Session, affected_id: int) -> AffectedPeriod:
    ap = db.get(AffectedPeriod, affected_id)
    if ap is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "找不到受影響節次")
    return ap


def _sub_out(sub: Substitution) -> SubstitutionOut:
    return SubstitutionOut(
        id=sub.id, affected_period_id=sub.affected_period_id, type=sub.type,
        type_label=SUBSTITUTION_TYPE_CN.get(sub.type, sub.type),
        handler_teacher_id=sub.handler_teacher_id,
        handler_name=sub.handler.name if sub.handler else None,
        counts_toward_hours=sub.counts_toward_hours, funding_source=sub.funding_source,
        swap_date=sub.swap_date, swap_period_name=sub.swap_period_name,
        swap_class_names=sub.swap_class_names, swap_subject_name=sub.swap_subject_name,
        created_by_name=sub.created_by_name,
    )


@router.get("/affected-periods/{affected_id}/recommendations", response_model=RecommendationOut)
def get_recommendations(
    affected_id: int, db: Session = Depends(get_db), _: User = Depends(editor)
):
    """這一節該找誰代:硬性過濾後排序,每位候選附理由。"""
    affected = _get_affected(db, affected_id)
    rec = recommender.recommend(db, affected)
    return RecommendationOut(
        affected_period_id=rec.affected_period_id,
        no_candidate_hint=rec.no_candidate_hint,
        candidates=[
            CandidateOut(
                teacher_id=c.teacher_id, teacher_name=c.teacher_name,
                same_subject=c.same_subject, at_school_that_day=c.at_school_that_day,
                sub_periods_this_month=c.sub_periods_this_month, reasons=list(c.reasons),
            )
            for c in rec.candidates
        ],
    )


@router.get("/affected-periods/{affected_id}/swap-options", response_model=SwapOptionsOut)
def get_swap_options(
    affected_id: int,
    teacher_id: int | None = Query(default=None),
    weeks: int = Query(default=swap_options.DEFAULT_WEEKS, ge=1, le=swap_options.MAX_WEEKS),
    db: Session = Depends(get_db),
    _: User = Depends(editor),
):
    """調課:列出請假那週起 `weeks` 週內可對調的節次。未指定教師時只找也教這個班的老師。"""
    affected = _get_affected(db, affected_id)
    return SwapOptionsOut.model_validate(
        asdict(swap_options.search(db, affected, teacher_id=teacher_id, weeks=weeks)))


@router.get("/swap-slips", response_model=SlipsOut)
def get_swap_slips(
    semester_id: int = Query(...),
    affected_period_ids: list[int] = Query(..., min_length=1, max_length=500),
    db: Session = Depends(get_db),
    _: User = Depends(editor),
):
    """調課通知單(教師調課單 + 班級調課單)的列印資料。非調課的節次略過。"""
    try:
        slips = swap_slips.build(db, semester_id, affected_period_ids)
    except SlipError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    return SlipsOut.model_validate(asdict(slips))


@router.get("/substitute-slips", response_model=SlipsOut)
def get_substitute_slips(
    semester_id: int = Query(...),
    affected_period_ids: list[int] = Query(..., min_length=1, max_length=500),
    db: Session = Depends(get_db),
    _: User = Depends(editor),
):
    """代課通知單(教師代課單 + 班級代課單)的列印資料。代課與併班以外的節次略過。"""
    try:
        slips = sub_slips.build(db, semester_id, affected_period_ids)
    except SlipError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, str(exc)) from exc
    return SlipsOut.model_validate(asdict(slips))


@router.put("/affected-periods/{affected_id}/substitution", response_model=SubstitutionOut)
def assign_substitution(
    affected_id: int,
    body: AssignRequest,
    db: Session = Depends(get_db),
    user: User = Depends(editor),
):
    """指派處置(代課/調課/併班/自習/不處理);指派即生效並通知處理教師。"""
    affected = _get_affected(db, affected_id)
    try:
        sub = sub_service.assign(
            db, affected,
            sub_type=body.type, handler_teacher_id=body.handler_teacher_id,
            counts_toward_hours=body.counts_toward_hours, funding_source=body.funding_source,
            swap_entry_id=body.swap_entry_id, swap_date=body.swap_date,
            swap_period_no=body.swap_period_no,
            created_by_user_id=user.id, created_by_name=user.username,
        )
    except sub_service.SubstitutionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc

    db.add(AuditLog(
        user_id=user.id, username=user.username, action="assign_substitution",
        target_type="affected_period", target_id=affected.id,
        detail=(
            f"{affected.leave_request.teacher.name} {affected.date} {affected.period_name}"
            f" → {SUBSTITUTION_TYPE_CN.get(sub.type, sub.type)}"
            + (f"({sub.handler.name})" if sub.handler else "")
        )[:500],
    ))
    db.commit()
    db.refresh(sub)
    return _sub_out(sub)


@router.delete("/affected-periods/{affected_id}/substitution", status_code=status.HTTP_200_OK)
def clear_substitution(
    affected_id: int, db: Session = Depends(get_db), user: User = Depends(editor)
):
    """撤回處置:退回待處理,已指派教師收到取消通知。"""
    affected = _get_affected(db, affected_id)
    try:
        sub_service.clear(db, affected, actor_name=user.username)
    except sub_service.SubstitutionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc)) from exc
    db.commit()
    return {"affected_period_id": affected.id, "status": affected.status}


@router.get("/substitution-types", response_model=dict[str, str])
def substitution_types(_: User = Depends(editor)):
    return {t.value: SUBSTITUTION_TYPE_CN[t.value] for t in SubstitutionType}


@router.get("/substitution-funding-sources", response_model=list[str])
def substitution_funding_sources(_: User = Depends(editor)):
    """代課鐘點的計費方式常用選項(印在代課通知單上);畫面也允許自行輸入。"""
    return list(FUNDING_SOURCES)


@router.get("/affected-periods/{affected_id}/substitution", response_model=SubstitutionOut | None)
def get_substitution(
    affected_id: int, db: Session = Depends(get_db), _: User = Depends(editor)
):
    sub = db.scalar(select(Substitution).where(Substitution.affected_period_id == affected_id))
    return _sub_out(sub) if sub else None
